include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -1278,6 +1286,110 @@ namespace internal
 
       const auto reference_cell = dof_handler_fine.get_fe(0).reference_cell();
 
+      std::unique_ptr<internal::MatrixFreeFunctions::HangingNodes<dim>>
+                                             hanging_nodes;
+      std::vector<std::vector<unsigned int>> lexicographic_mappings;
+
+      {
+        // algorithm is only needed on active levels
+        bool use_fast_hanging_node_algorithm =
+          mg_level_coarse == numbers::invalid_unsigned_int;
+
+        const auto &tria = dof_handler_coarse.get_triangulation();
+
+        // algorithm can be only used on meshes consisting of hypercube and
+        // simplices
+        if (use_fast_hanging_node_algorithm)
+          {
+            const auto &reference_cells = tria.get_reference_cells();
+            use_fast_hanging_node_algorithm =
+              std::all_of(reference_cells.begin(),
+                          reference_cells.end(),
+                          [](const auto &r) {
+                            return r.is_hyper_cube() || r.is_simplex();
+                          });
+          }
+
+        // local p-refinement is not supported
+        if (use_fast_hanging_node_algorithm)
+          {
+            const auto &fes = dof_handler_coarse.get_fe_collection();
+
+            use_fast_hanging_node_algorithm &=
+              std::all_of(fes.begin(), fes.end(), [&fes](const auto &fe) {
+                return fes[0].compare_for_domination(fe) ==
+                       FiniteElementDomination::Domination::
+                         either_element_can_dominate;
+              });
+          }
+
+        // check that all components are either supported or not
+        if (use_fast_hanging_node_algorithm)
+          {
+            const auto supported_components = internal::MatrixFreeFunctions::
+              HangingNodes<dim>::compute_supported_components(
+                dof_handler_coarse.get_fe_collection());
+
+            use_fast_hanging_node_algorithm &=
+              std::any_of(supported_components.begin(),
+                          supported_components.end(),
+                          [](const auto &supported_components) {
+                            return std::all_of(supported_components.begin(),
+                                               supported_components.end(),
+                                               [](const auto &a) {
+                                                 return a == true;
+                                               });
+                          });
```

Comment:
I think this would be easier to understand if the type of `supported_components` were spelled out (`std::vector<std::vector<bool>>`) or if the inner variable were called something different `[](const std::vector<bool>&supported_components_per_fe)`. Or is it `..._per_base_element`?